### PR TITLE
Function Calls for MethodValues (no Interpreter support)

### PR DESCRIPTION
**Pull request by Krovatkin**
This PR adds infrastructure for adding function calls for `MethodValue`s. Interpreter doesn't know how to handle function calls yet.

### CODE REVIEW

File: c10/util/StringUtil.h [hunk 1]
```diff
@@ -61,6 +61,19 @@ inline std::string Join(const std::string& delimiter, const Container& v) {
   return s.str();
 }
 
+
+inline std::vector<std::string> Split(const std::string& input, const std::string& delimiter) {
```
Comment (Krovatkin) [inline c10/util/StringUtil.h]: There seems to be many different implementations of `Split` sprinkled all over the place. I'm not sure if we might want to consider unifying/merging them. Neither am I sure if this work needs to be a part of this PR.

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -563,6 +564,55 @@ struct to_ir {
     ConstantPooling(to_clean);
   }
 
+void runInline(Block* block) {
+
+  Node* cur = block->nodes().front();
+  Node* end = block->return_node();
+
+  while (cur != end)
```
Comment (Krovatkin) [inline torch/csrc/jit/script/compiler.cpp]: I did consider iterators, but since we need to keep track of several iteration points, it looked rather clunky.

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -2102,6 +2216,81 @@ struct to_ir {
...
+        }
+        case '.':
+        {
+          auto select = Select(expr);
+          auto suv = getFullyQualifiedName(select.value(), names);
+          names.push_back(select.selector().name());
+          return (suv) ? suv->attr(select.range(), method, select.selector().name()) : nullptr;
+          break;
+        }
+        default:
```
Comment (Krovatkin) [inline torch/csrc/jit/script/compiler.cpp]: This PR doesn't support expressions like `modules[i].method1()` or anything else except for selectors for now 😢

File: torch/csrc/jit/script/compiler.cpp [hunk 4]
```diff
@@ -563,6 +564,55 @@ struct to_ir {
...
+      auto new_output = inlineCallTo(*cur->owningGraph(), *method.graph().get(), inputs).at(0);
+      if (old_output.at(0)->hasUniqueName())
+      {
+        auto name = old_output.at(0)->uniqueName();
+        new_output->setUniqueName(name);
+      }
+
+      old_output.at(0)->replaceAllUsesWith(new_output);
+      next = cur->next();
+      cur->destroy();
```
Comment (Krovatkin) [inline torch/csrc/jit/script/compiler.cpp]: DCE won't touch this, since it's marked as having side effects, so we have to remove it manually.

File: torch/csrc/jit/script/compiler.cpp [hunk 5]
```diff
@@ -2035,6 +2085,70 @@ struct to_ir {
     } else {
       auto inputs = getNamedValues(apply.inputs(), true);
       auto attributes = emitAttributes(apply.attributes());
+
+      auto r = getFullyQualifiedNameAndSchema(apply.callee(), inputs, attributes);
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: I do not think this is the right approach. It replicates a bunch of the SugaredValue behavior locally just to try to get the fully-qualified name for a function relative to the module root. MethodValue should just track its fully qualified name itself. I would not expect to see many changes in compiler.cpp

File: torch/csrc/jit/script/compiler.cpp [hunk 6]
```diff
@@ -563,6 +564,55 @@ struct to_ir {
     ConstantPooling(to_clean);
   }
 
+void runInline(Block* block) {
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: Separate file please. compiler.cpp is long enough. This will need to be detangled from things like the environment. Otherwise the PR is not really accomplishing its purpose.

**Review by zdevito — COMMENTED** [review]
Hi, I added some high-level comments. Let's chat more about how we will name functions, etc.

File: aten/src/ATen/core/jit_type.h [hunk 7]
```diff
@@ -883,6 +884,31 @@ struct CAFFE2_API StringType : public Type {
   : Type(TypeKind::StringType) {}
 };
 
+struct FunctionType;
+using FunctionTypePtr = std::shared_ptr<FunctionType>;
+// This type represents a Python Function
+struct CAFFE2_API FunctionType : public Type {
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply aten/src/ATen/core/jit_type.h]: given that we are inlining right away we could s tart with a blank `FunctionType` and add more members lazily.

File: c10/util/StringUtil.h (hunk 1 above)
Comment (Krovatkin) [reply c10/util/StringUtil.h]: gut

File: aten/src/ATen/core/jit_type.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (zdevito) [reply aten/src/ATen/core/jit_type.h]: The simplest thing to do right now might be to make the FunctionType object contain the std::shared_ptr<script::Function> object. That way we can still use prim::Constant, but it doesn't need any attributes. That should be an easy tweak to what is here now.

File: torch/csrc/jit/passes/inliner.h [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#pragma once
+
+#include <torch/csrc/jit/ir.h>
+
+namespace torch {
+namespace jit {
+
+TORCH_API void Inline(Block* block, bool recurse = true);
```
Comment (zdevito) [inline torch/csrc/jit/passes/inliner.h]: Can you comment in the code why the recurse flag exists? (not what it is but why it is)

File: torch/csrc/jit/passes/lower_tuples.cpp [hunk 9]
```diff
@@ -170,9 +169,18 @@ void LowerAllTuples(std::shared_ptr<Graph>& graph) {
   EnsureNoTuples(graph->block());
 }
 
+void flattenPrintInputs(Node* n) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/lower_tuples.cpp]: I am not sure what this is. Was this a bad merge? Doesn't seem related to this PR.

File: torch/csrc/jit/script/compiler.cpp [hunk 10]
```diff
@@ -2108,6 +2109,68 @@ struct to_ir {
     } else {
       auto inputs = getNamedValues(apply.inputs(), true);
       auto attributes = emitAttributes(apply.attributes());
+
+      if (auto mv = std::dynamic_pointer_cast<MethodValue>(sv)) {
+        auto& callee = mv->getMethod();
+        try {
+          callee.ensure_defined();
+        } catch (std::exception&) {
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: It is important to catch the specific exception. Otherwise this will hide all sorts of bad errors and report them as method recursion.

File: torch/csrc/jit/script/compiler.cpp [hunk 11]
```diff
@@ -2108,6 +2109,68 @@ struct to_ir {
...
+
+        //add fun_constant
+        call_node->addInput(fun_constant);
+        auto& matched_inputs = matched_schema.value().inputs;
+        for (auto input : matched_inputs)
+        {
+          call_node->addInput(input);
+        }
+
+        call_node->eraseOutput(0);
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: Following previous convention, I think it would be better if we added an `insertCallFunction(Value * callee, std::vector<Value*> inputs)` call to Graph. Like others, it should never do user-facing error reporting, but it should handle the details of what a CallFunction node looks like.

File: torch/csrc/jit/script/sugared_value.cpp [hunk 12]
```diff
@@ -28,15 +28,6 @@ std::shared_ptr<SugaredValue> PrintValue::call(
   // print(a, b) is treated as a (a, b) tuple input.
 
   std::vector<Value*> lowered_inputs = toValues(*m.graph(), inputs);
-  if (lowered_inputs.size() == 1 &&
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: Ok, I see the move of this code into the lower tuples pass. However, I don't think it does the same thing, because this only lowers the tuple if it is the only argument to prim::Print, while the other one appears to do it universally.

File: torch/csrc/jit/script/compiler.cpp [hunk 13]
```diff
@@ -2108,6 +2109,68 @@ struct to_ir {
     } else {
       auto inputs = getNamedValues(apply.inputs(), true);
       auto attributes = emitAttributes(apply.attributes());
+
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: This code should be in MethodValue::call and not inline in the compiler here.

File: torch/csrc/jit/script/sugared_value.h [hunk 14]
```diff
@@ -226,6 +226,13 @@ struct MethodValue : public SugaredValue {
         method.emit_call(graph, loc, inputs, attributes));
   }
 
+  Function& getMethod() {
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: These shouldn't be necessary, MethodValue should handle its own call logic.

Comment (zdevito) [reply torch/csrc/jit/script/sugared_value.h]: Also the current MethodValue::call is dead.

**Review by zdevito — COMMENTED** [review]
Nice progress. This is getting close, but there are a few restructuring things I suggest:

* FunctionType can simply contain the script::Function* object
* We do not need the ivalue object in this PR, we can just directly create prim::Constant with a FunctionType return type.
* Logic for emitting the CallFunction should move into Graph::createCall, and the actual use of that function should be done in MethodValue::call. There is no need to put a special case in the compiler (and in fact, one reason for wanting first-class calls is to remove special casing in the compiler).

File: torch/csrc/jit/passes/dead_code_elimination.cpp [hunk 15]
```diff
@@ -107,6 +108,12 @@ class DeadCodeEliminator {
       if (hasSideEffects(node)) {
         mark(node);
       }
+
```
Comment (zdevito) [inline torch/csrc/jit/passes/dead_code_elimination.cpp]: Passes don't typically descend into subgraphs. Is this still needed now that the Function* is actually stored in the type?

File: torch/csrc/jit/passes/inliner.cpp [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+      {
+        for (auto ib : it->blocks())
+        {
+            removeFunctionConstants(ib);
+        }
+      }
+    }
+  }
+
+  void unifyIfs(Block* b)
```
Comment (zdevito) [inline torch/csrc/jit/passes/inliner.cpp]: Why is this here? It is breaking loose coupling for inlining to need to think about how it would affect other statements.

File: torch/csrc/jit/passes/inliner.cpp [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+        if (old_output.at(0)->hasUniqueName()) {
+          auto name = old_output.at(0)->uniqueName();
+          new_output->setUniqueName(name);
+        }
+
+        old_output.at(0)->replaceAllUsesWith(new_output);
+        next = cur->next();
+        cur->destroy();
+      } else {
+        if (cur->hasAttribute(attr::Subgraph)) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/inliner.cpp]: Optimization of subgraphs should be handled by the thing inserting the subgraph. Most of our passes do not descend into subgraphs, which is why this code needs to keep modifying passes. I don't want to start mixing typical behavior and have some passes look at subgraphs but others that do not.

File: torch/csrc/jit/script/compiler.cpp [hunk 18]
```diff
@@ -2391,6 +2392,7 @@ struct to_ir {
       body_block->registerOutput(fn_simple_output);
       node_output = fork_node->output()->setType(
           FutureType::create(fn_simple_output->type()));
+      Inline(body_block);
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: Why? runCleanupPasses will run on the subgraph below and do the inlining.

File: torch/csrc/jit/script/sugared_value.cpp [hunk 19]
```diff
@@ -14,6 +14,81 @@ struct NoneValue : SugaredValue {
...
+    throw ErrorReport(loc)
+        << " method '" << callee.name()
+        << "' is called recursively involving this call site. Recursive calls are not supported";
+  }
+
+  auto callee_graph = callee.graph();
+  auto& schema = callee.getSchema();
+
+  std::vector<NamedValue> inputsWithSelf(inputs.begin(), inputs.end());
+  auto self_val = self_.value(*caller.graph());
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: Self values are always present. This was refactored a few weeks ago.

File: torch/csrc/jit/script/sugared_value.cpp [hunk 20]
```diff
@@ -14,6 +14,81 @@ struct NoneValue : SugaredValue {
...
+        << " mismatch between arguments and parameters " << failure_messages.str();
+  }
+
+  if (matched_schema.value().return_types.size() > 1)
+  {
+    throw ErrorReport(loc)
+        << "multiple returns aren't yet supported";
+  }
+
+  auto fun_constant = graph.create(prim::Constant);
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: Code to correctly construct CallFunction should be in ir.h:

```
::createFunctionCall(Value* fn, at::ArrayRef inputs);
```

File: torch/csrc/jit/script/sugared_value.cpp (hunk 19 above)
Comment (zdevito) [reply torch/csrc/jit/script/sugared_value.cpp]: Also, this does not replace FunctionValue::call which the root place where function calls get inserted. MethodValue should be unmodified in this patch. This code should be in FunctionValue::call

Comment (zdevito) [reply torch/csrc/jit/script/sugared_value.cpp]: This also seems like the wrong place to put this code. We have a method on `Function` with the name `emit_call_to`. I feel like this implementation should be there. This also explains why this code seems so long: it is copying all of the logic that already exists in that function to a new place without deleting it.

**Review by zdevito — COMMENTED** [review]
Inline comments on a few places where this did not merge nicely with the current codebase.

File: torch/csrc/jit/script/compiler.cpp (hunk 18 above)
Comment (Krovatkin) [reply torch/csrc/jit/script/compiler.cpp]: `lambdaLiftFork` does sort of DCE i.e. it only lifts values that are used in a subgraph. If we leave a function call in place we end up passing in more values than really necessary and when we recompile the same function in `emitModuleHook` which will contain an already inlined subgraph we end up with different graphs. It's actually a very common theme with all these test failures. We *can't* do inlining early enough to not hinder some of optimizations so we have to re-do them.

File: torch/csrc/jit/script/sugared_value.cpp (hunk 19 above)
Comment (Krovatkin) [reply torch/csrc/jit/script/sugared_value.cpp]: okay... let me see if I can move stuff around

File: torch/csrc/jit/script/sugared_value.cpp (hunk 20 above)
Comment (Krovatkin) [reply torch/csrc/jit/script/sugared_value.cpp]: will fix

File: torch/csrc/jit/passes/inliner.cpp (hunk 16 above)
Comment (Krovatkin) [reply torch/csrc/jit/passes/inliner.cpp]: type unification is typically done by `compiler.cpp` which now doesn't have the full inlined graph (hence, no precise types) since we inline later and even later in `emitModuleHook` we compile the inlined graph which now will have more precise types, so we end up with different graphs again.

File: torch/csrc/jit/passes/inliner.cpp (hunk 17 above)
Comment (Krovatkin) [reply torch/csrc/jit/passes/inliner.cpp]: I think that inliner is a **very** special case. It is more similar to `compiler.cpp` rather than a regular optimization. Let me check if making `compiler.cpp` calling `Inline` is good enough. I think it was still breaking some tests if `Inliner` didn't expect subgraphs.

File: torch/csrc/jit/passes/dead_code_elimination.cpp (hunk 15 above)
Comment (Krovatkin) [reply torch/csrc/jit/passes/dead_code_elimination.cpp]: actually, I think I might be able to remove this particular one, but then Inliner will need to descend into subgraphs and clean up unused `Function` literals

File: torch/csrc/jit/script/compiler.cpp (hunk 18 above)
Comment (Krovatkin) [reply torch/csrc/jit/script/compiler.cpp]: i.e. running it on a lifted subgraph is too late.

**Comment by Krovatkin** [comment]
Closing in favour of #21052
